File: src/pipeline/rpm-source.js
// rpm-source — interface única pra fontes de RPM (e TPS).
//
// O Shift Light precisa de RPM em tempo real, mas a fonte muda conforme
// o setup do piloto: ECU via Bluetooth (Injepro/MoTeC), OBD-II, simulador,
// arquivo replay. Este módulo isola essa diversidade atrás de um contrato
// estável que o `shift-light-bridge` consome.
//
// Contrato de RpmSource:
//   {
//     getRpm():    number | null     // RPM corrente, null se sem leitura fresca
//     getTps():    number | null     // 0..100, null se sensor ausente
//     getStatus(): 'connected' | 'degraded' | 'lost'
//     start():     Promise<void> | void
//     stop():      void
//   }

const FRESHNESS_MS_DEFAULT = 500;

// createManualRpmSource — caller atualiza via setRpm()/setTps().
// Útil pra testes, simulador, ou pra ligar manualmente a qualquer canal
// externo (Web Bluetooth, MQTT, WebSocket) que ainda não vira adapter.
export function createManualRpmSource({ freshnessMs = FRESHNESS_MS_DEFAULT, now = () => Date.now() } = {}) {
  let rpm = null, tps = null, lastUpdate = -Infinity, started = false;
  return {
    setRpm(v) { rpm = Number.isFinite(v) && v >= 0 ? v : null; lastUpdate = now(); },
    setTps(v) { tps = Number.isFinite(v) ? v : null; },
    getRpm() {
      if (!started) return null;
      if (now() - lastUpdate > freshnessMs) return null;
      return rpm;
    },
    getTps() { return started ? tps : null; },
    getStatus() {
      if (!started) return 'lost';
      const age = now() - lastUpdate;
      if (age > freshnessMs * 2) return 'lost';
      if (age > freshnessMs) return 'degraded';
      return 'connected';
    },
    start() { started = true; },
    stop()  { started = false; rpm = null; tps = null; }
  };
}

// createMockRpmSource — RPM "scriptado" por trajetória (testes E2E).
// Aceita uma função `(t) => { rpm, tps }` que define o que reportar.
export function createMockRpmSource(scriptFn) {
  if (typeof scriptFn !== 'function') {
    throw new Error('createMockRpmSource: scriptFn obrigatório');
  }
  let started = false, lastT = null;
  return {
    tick(t) { lastT = t; },
    getRpm() {
      if (!started || lastT == null) return null;
      const v = scriptFn(lastT);
      return v && Number.isFinite(v.rpm) ? v.rpm : null;
    },
    getTps() {
      if (!started || lastT == null) return null;
      const v = scriptFn(lastT);
      return v && Number.isFinite(v.tps) ? v.tps : null;
    },
    getStatus() { return started ? 'connected' : 'lost'; },
    start() { started = true; },
    stop()  { started = false; lastT = null; }
  };
}

// connectBleRpmSource — esqueleto pra Web Bluetooth (Injepro/OBD-II).
// O parsing de characteristic é específico do dispositivo; o adapter
// concreto fica no app. Aqui só garantimos a forma do contrato.
export function createBleRpmSource({
  device,             // BluetoothDevice (já pareado pelo caller)
  rpmCharacteristic,  // BluetoothRemoteGATTCharacteristic
  parseValue,         // (DataView) → { rpm, tps }
  freshnessMs = FRESHNESS_MS_DEFAULT,
  now = () => Date.now()
} = {}) {
  if (!device || !rpmCharacteristic || typeof parseValue !== 'function') {
    throw new Error('createBleRpmSource: device + rpmCharacteristic + parseValue obrigatórios');
  }
  let rpm = null, tps = null, lastUpdate = -Infinity, started = false;
  function onValueChanged(ev) {
    const v = ev?.target?.value;
    if (!v) return;
    let parsed;
    try { parsed = parseValue(v); } catch { parsed = null; }
    if (!parsed) return;
    if (Number.isFinite(parsed.rpm)) rpm = parsed.rpm;
    if (Number.isFinite(parsed.tps)) tps = parsed.tps;
    lastUpdate = now();
  }
  return {
    async start() {
      if (started) return;
      rpmCharacteristic.addEventListener('characteristicvaluechanged', onValueChanged);
      await rpmCharacteristic.startNotifications();
      started = true;
    },
    stop() {
      if (!started) return;
      rpmCharacteristic.removeEventListener('characteristicvaluechanged', onValueChanged);
      try { rpmCharacteristic.stopNotifications(); } catch { /* ignore */ }
      started = false;
      rpm = null; tps = null;
    },
    getRpm() {
      if (!started) return null;
      if (now() - lastUpdate > freshnessMs) return null;
      return rpm;
    },
    getTps() { return started ? tps : null; },
    getStatus() {
      if (!started || !device.gatt?.connected) return 'lost';
      const age = now() - lastUpdate;
      if (age > freshnessMs * 2) return 'lost';
      if (age > freshnessMs) return 'degraded';
      return 'connected';
    }
  };
}

File: ios/p1fast-core/Sources/P1FastCore/RpmSource.swift
// ═══════════════════════════════════════════════════════════
// RpmSource — port Swift de src/pipeline/rpm-source.js
// ═══════════════════════════════════════════════════════════
// Interface única para fontes de RPM (e TPS). Isolam a diversidade
// de fontes (Manual / Mock / BLE-T4000) atrás de um contrato estável
// consumido pelo `ShiftLightBridge` (MS-8.5).
//
// Esta release contém:
//   • RpmSource — protocolo
//   • ManualRpmSource — caller seta manualmente (testes / Web Bluetooth
//     externo / qualquer canal que ainda não vire adapter)
//   • MockRpmSource — RPM "scriptado" por trajetória (testes E2E)
//
// O adapter BLE T4000 (CoreBluetooth real) entra em MS-9, junto com
// `gear_signatures` / `gear_ratios` / `dyno_curve` cadastrados.
//
// MS-8.6 (port 2026-05-04).

import Foundation

public enum RpmSourceConst {
    public static let freshnessMsDefault: Double = 500
}

public enum RpmSourceStatus: String, Sendable, Codable {
    case connected
    case degraded
    case lost
}

public protocol RpmSource: AnyObject, Sendable {
    /// RPM corrente, nil se sem leitura fresca.
    func getRpm() -> Double?
    /// 0..100, nil se sensor ausente.
    func getTps() -> Double?
    /// Status de saúde do feed.
    func getStatus() -> RpmSourceStatus
    /// Inicia o feed (em BLE roda async; aqui sync por simplicidade).
    func start()
    /// Para o feed e limpa estado.
    func stop()
}

/// Fonte manual — caller atualiza via setRpm/setTps. Útil pra testes,
/// simulador ou pra ligar manualmente a qualquer canal externo (Web
/// Bluetooth, MQTT, WebSocket) que ainda não tenha adapter dedicado.
public final class ManualRpmSource: RpmSource, @unchecked Sendable {
    private var rpm: Double? = nil
    private var tps: Double? = nil
    private var lastUpdate: Double = -.infinity
    private var started: Bool = false
    private let freshnessMs: Double
    private let now: () -> Double

    public init(
        freshnessMs: Double = RpmSourceConst.freshnessMsDefault,
        now: @escaping () -> Double = { Date().timeIntervalSince1970 * 1000 }
    ) {
        self.freshnessMs = freshnessMs
        self.now = now
    }

    public func setRpm(_ v: Double?) {
        if let v, v.isFinite, v >= 0 {
            rpm = v
        } else {
            rpm = nil
        }
        lastUpdate = now()
    }

    public func setTps(_ v: Double?) {
        if let v, v.isFinite { tps = v } else { tps = nil }
    }

    public func getRpm() -> Double? {
        guard started else { return nil }
        if now() - lastUpdate > freshnessMs { return nil }
        return rpm
    }

    public func getTps() -> Double? {
        started ? tps : nil
    }

    public func getStatus() -> RpmSourceStatus {
        guard started else { return .lost }
        let age = now() - lastUpdate
        if age > freshnessMs * 2 { return .lost }
        if age > freshnessMs     { return .degraded }
        return .connected
    }

    public func start() { started = true }
    public func stop()  { started = false; rpm = nil; tps = nil }
}

/// Fonte mock — RPM scriptado por uma função `(t) -> (rpm, tps)`.
/// Use com `tick(t)` antes de cada leitura para definir o instante.
public final class MockRpmSource: RpmSource, @unchecked Sendable {
    public typealias Script = (Double) -> (rpm: Double?, tps: Double?)

    private let script: Script
    private var started: Bool = false
    private var lastT: Double? = nil

    public init(script: @escaping Script) {
        self.script = script
    }

    public func tick(_ t: Double) { lastT = t }

    public func getRpm() -> Double? {
        guard started, let t = lastT else { return nil }
        let v = script(t)
        guard let r = v.rpm, r.isFinite else { return nil }
        return r
    }

    public func getTps() -> Double? {
        guard started, let t = lastT else { return nil }
        let v = script(t)
        guard let p = v.tps, p.isFinite else { return nil }
        return p
    }

    public func getStatus() -> RpmSourceStatus {
        started ? .connected : .lost
    }

    public func start() { started = true }
    public func stop()  { started = false; lastT = nil }
}
